File: src/main/java/net/onelitefeather/vulpes/generator/util/BranchFilter.java
package net.onelitefeather.vulpes.generator.util;

import net.onelitefeather.vulpes.generator.domain.configuration.BranchFilterConfiguration;
import org.jetbrains.annotations.ApiStatus;

import java.util.List;

@ApiStatus.Internal
public final class BranchFilter {

    /**
     * Filters the branches and returns a list with the filtered branches.
     *
     * @param branches            the branches which should be filtered
     * @param filterConfiguration the filter configuration to use
     * @return a list with the filtered branches
     */
    public static List<String> filterBranches(List<String> branches, BranchFilterConfiguration filterConfiguration) {
        return branches.stream().filter(entry ->
                        filterConfiguration.exclude().stream()
                                .noneMatch(entry::startsWith)
                )
                .toList();
    }

    private BranchFilter() {
        throw new UnsupportedOperationException("This class cannot be instantiated");
    }
}

File: src/main/kotlin/net/theevilreaper/vulpes/generator/util/BranchFilter.kt
package net.theevilreaper.vulpes.generator.util

/**
 * The class contains some utility methods for the branch filtering.
 * @author theEvilReaper
 * @version 1.0.0
 * @since
 **/
object BranchFilter {

    /**
     * Filters the branches and returns a list with the filtered branches.
     * @param branches The branches which should be filtered
     * @param predicate The predicate which should be used for the filtering
     * @return A list with the filtered branches
     */
    inline fun filterBranches(branches: List<String>, crossinline predicate: (String) -> Boolean): List<String> {
        if (branches.isEmpty()) return emptyList()
        return branches.filter { predicate(it) }.toList()
    }
}
